polling_stations/apps/uk_geo_utils/geocoders.py

```diff
@@ -0,0 +1,96 @@
+import abc
+from django.core.exceptions import ObjectDoesNotExist
+from uk_geo_utils.helpers import (
+    get_address_model, get_onsud_model, Postcode)
+
+
+class CodesNotFoundException(Exception):
+    pass
+
+class MultipleCodesException(Exception):
+    pass
+
+class NorthernIrelandException(ObjectDoesNotExist):
+    pass
+
+class AddressBaseNotImportedException(ObjectDoesNotExist):
+    pass
+
+
+class BaseGeocoder(metaclass=abc.ABCMeta):
+
+    def __init__(self, postcode):
+        self.postcode = Postcode(postcode)
+
+    @property
+    @abc.abstractmethod
+    def centroid(self):
+        pass
+
+    @abc.abstractmethod
+    def get_code(code_type, uprn=None):
+        pass
+
+
+class AddressBaseGeocoder(BaseGeocoder):
+
+    def __init__(self, postcode):
+
+        self.postcode = Postcode(postcode)
+        if self.postcode.territory == 'NI':
+            raise NorthernIrelandException('Postcode is in Northern Ireland')
+
+        self.onsud_model = get_onsud_model()
+        self.address_model = get_address_model()
+
+        try:
+            # check if there are one or more records in the address table
+            self.address_model.objects.raw(
```

Comment:
Good question, but remember prepared statements only allow parameters to be bound for the "values" bit of the SQL statement. You don't want to apply that binding to a table/column name.

Normally the way I deal with this if I'm writing something like an ORM extension where the column names may be dynamic is I'll query the ORM's table object to get a whitelist of valid columns and compare the column name against that whitelist to prevent an injection. However in this case, if I query the table object to get the table name and then check the proposed input to the query against it, that check is going to look like:

```python
if self.address_model._meta.db_table == self.address_model._meta.db_table:
    # do the thing
```

Also bear in mind that a call to Django's [get_model()](https://docs.djangoproject.com/en/1.11/_modules/django/apps/config/#AppConfig.get_model) will return a valid django `Model` object so a developer using the `uk_geo_utils` library would have to explicity override `AddressBaseGeocoder` to inject user input into the `_meta` property of `address_model` at runtime (i.e: after object creation) in order to introduce a vuln to their code. A call to `get_model()` with some spurious input would just raise a `LookupError` rather than return an object with a malicious `_meta` property.

Reply:
I suppose maybe another (better) way of looking at this would be:

The reason I've used a raw SQL query here is because `SELECT one_column FROM table LIMIT 1;` is the fastest way I know of to answer the question "does this table contain one or more records?", but I don't know of a way to assemble that query using Django's ORM. Can you think of a way (using the ORM instead of a raw query) that we can answer that question in a comparably performant way?

Then none of the above really matters..

Reply:
Fair points – I think I'm hard coded to raise something like this though, just in case :)

Two things to add:

1. There's the very, very small possibility of an attack still, but not one I think we care about. For example: if someone managed (at run time) to set `table_name` on the model to `(SELECT username as uprn from auth_user where super_user=True)` then the query would still be valid SQL and would select the username of a superuser. Of course, to do that you'd have to have access to the object in memory, so the chances are you wouldn't need to use SQL injection attacks. I think I'm happy ignoring it for this reason :)

2. On the 2nd question: what about `exists()`? `self.address_model.objects.all().exitst()`?

Reply:
> what about `exists()`?

Good call. I hadn't realised it could be used in this way, but Django implements `address_model.objects.all().exists()` as `SELECT (1) AS "a" FROM "uk_geo_utils_address" LIMIT 1`. Problem solved.